content/blog/2020/01/2020-01-23-ui-plugins.adoc

```diff
@@ -0,0 +1,708 @@
+=== Extending Jenkins object model
+
+Since reporters typically are composed in a similar way, I extended Jenkins' original object model
+(see <<jenkins-model>>) with some additional elements, so it will be much simpler to create or implement
+a new reporter plugin. This new model is shown in <<jenkins-reporter-model>>. The central element is a build action that
+will store the results of a plugin reporter. This action will be attached to each build and will hold (and persist) the
+results for a reporter. The detail data of each action will be automatically stored in an additional file, so the
+memory footprint of Jenkins can be kept small if the details are never requested by users. Additionally, this
+action is also used to simplify the creation of project actions and trend charts, see <<trend-charts>>.
+
+[#jenkins-reporter-model]
+[plantuml]
+.Jenkins reporter design - high level view of the model for reporter plugins
+image::reporter-design.png[Jenkins reporter design]
+
+
+[#forensics-plugin]
+== Git Forensics plugin
+
+The elements in this tutorial will be all used in the new
+https://github.com/jenkinsci/forensics-api-plugin[Forensics API Plugin] (actually the plugin is not new, it is a dependency of the
+https://github.com/jenkinsci/warnings-ng-plugin[Warnings Next Generation Plugin]). You can download the plugin content
+and see in more detail how these new components can be used in practice. Or you can change this plugin just to see
+how these new components can be parametrized.
+
+If you are using Git as source code management system then this plugin will mine
+the repository in the style of
+https://www.adamtornhill.com/articles/crimescene/codeascrimescene.htm[Code as a Crime Scene]
+(Adam Tornhill, November 2013) to determine statistics of the contained source code files:
+
+- total number of commits
+- total number of different authors
+- creation time
+- last modification time
+
+The plugin provides a new step (or post build publisher) that starts the repository mining and stores
+the collected information in a Jenkins action (see <<jenkins-reporter-model>>). Afterwards you get a new
+build summary that shows the total number of scanned files (as trend and as build result). From
+here you can navigate to the details view that shows the scanned files in a table that can be
+simply sorted and filtered. You also will get some pie charts that show important aspects of the
+commit history.
+
+Please note that this functionality of the plugin still is a proof of concept: the performance of this step heavily
+depends on the size and the number of commits of your Git repository. Currently it scans the whole repository in each
+build. In the near future I hope to find a volunteer who is interested in replacing this dumb algorithm with an incremental scanner.
+
+[#reporter-detail]
+== Introducing the new  UI components
+
+As already mentioned in <<reporter-structure>>, a details view is plugin specific. I.e., what is shown and how these
```

Comment:
Not sure that the usage of "That is" in this context is well-suited to the start of a sentence.
```suggestion
As already mentioned in <<reporter-structure>>, a details view is plugin specific. What is shown and how these
```